File: apps/web/lib/ai/nlp.ts
export interface ParsedCommand {
  type: 'create_task' | 'complete_task' | 'navigate' | 'query' | 'schedule' | 'unknown'
  confidence: number
  task?: {
    title: string
    dueDate?: string
    priority?: 'low' | 'medium' | 'high'
    estimatedMinutes?: number
    recurrence?: string
    tags?: string[]
  }
  navigation?: string
  query?: {
    entity?: string
    timeframe?: string
    intent?: string
  }
  raw: string
}

function extractDate(text: string): string | undefined {
  const now = new Date()
  const today = now.toISOString().split('T')[0]

  const tomorrow = new Date(now)
  tomorrow.setDate(tomorrow.getDate() + 1)
  const tomorrowStr = tomorrow.toISOString().split('T')[0]

  const dayNames: Record<string, number> = {
    sunday: 0, monday: 1, tuesday: 2, wednesday: 3, thursday: 4, friday: 5, saturday: 6,
    sun: 0, mon: 1, tue: 2, wed: 3, thu: 4, fri: 5, sat: 6,
  }

  if (/\btoday\b/i.test(text)) return today
  if (/\btomorrow\b/i.test(text)) return tomorrowStr
  if (/\b(tonight|this evening)\b/i.test(text)) return today
  if (/\b(this\s+)?weekend\b/i.test(text)) {
    const saturday = new Date(now)
    saturday.setDate(saturday.getDate() + ((6 - saturday.getDay() + 7) % 7))
    return saturday.toISOString().split('T')[0]
  }

  for (const [name, dayIndex] of Object.entries(dayNames)) {
    if (new RegExp(`\\b${name}\\b`, 'i').test(text)) {
      const isNext = /\bnext\b/i.test(text)
      const target = new Date(now)
      const daysUntil = ((dayIndex - target.getDay() + 7) % 7) || (isNext ? 7 : 0)
      target.setDate(target.getDate() + daysUntil + (isNext && daysUntil === 0 ? 7 : 0))
      return target.toISOString().split('T')[0]
    }
  }

  const dateMatch = text.match(/\b(\d{1,2})\/(\d{1,2})(?:\/(\d{2,4}))?\b/)
  if (dateMatch) {
    const [, m, d, y] = dateMatch
    const year = y ? (y.length === 2 ? `20${y}` : y) : String(now.getFullYear())
    return `${year}-${m.padStart(2, '0')}-${d.padStart(2, '0')}`
  }

  const nextPatterns = [
    { pattern: /\bnext\s+week\b/i, days: 7 },
    { pattern: /\bnext\s+month\b/i, days: 30 },
    { pattern: /\b(in\s+)?a\s+week\b/i, days: 7 },
    { pattern: /\b(in\s+)?a\s+month\b/i, days: 30 },
  ]
  for (const { pattern, days } of nextPatterns) {
    if (pattern.test(text)) {
      const target = new Date(now)
      target.setDate(target.getDate() + days)
      return target.toISOString().split('T')[0]
    }
  }

  const relMatch = text.match(/\b(in\s+(\d+)\s+(day|days|week|weeks))\b/i)
  if (relMatch) {
    const num = parseInt(relMatch[2])
    const unit = relMatch[3].toLowerCase()
    const target = new Date(now)
    if (unit.startsWith('week')) target.setDate(target.getDate() + num * 7)
    else target.setDate(target.getDate() + num)
    return target.toISOString().split('T')[0]
  }

  return undefined
}

function extractRecurrence(text: string): string | undefined {
  const patterns: Array<[RegExp, string]> = [
    [/\b(every|each)\s+day\b/i, 'daily'],
    [/\b(every|each)\s+week\b/i, 'weekly'],
    [/\b(every|each)\s+month\b/i, 'monthly'],
    [/\b(every|each)\s+(monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b/i, 'weekly'],
    [/\b(weekdays|week days|every weekday)\b/i, 'weekdays'],
    [/\b(daily|everyday|every day)\b/i, 'daily'],
    [/\b(weekly|every week)\b/i, 'weekly'],
    [/\b(monthly|every month)\b/i, 'monthly'],
  ]
  for (const [pattern, value] of patterns) {
    if (pattern.test(text)) return value
  }
  return undefined
}

function extractPriority(text: string): 'low' | 'medium' | 'high' | undefined {
  if (/\b(urgent|critical|asap|high priority|important|p0|p1)\b/i.test(text)) return 'high'
  if (/\b(low priority|whenever|someday|optional|p3|p4)\b/i.test(text)) return 'low'
  return undefined
}

function extractMinutes(text: string): number | undefined {
  const match = text.match(/\b(\d+)\s*(min|minute|minutes|hour|hours|hr|hrs)\b/i)
  if (!match) return undefined
  const num = parseInt(match[1])
  const unit = match[2].toLowerCase()
  if (unit.startsWith('hour') || unit.startsWith('hr')) return num * 60
  return num
}

function extractTags(text: string): string[] | undefined {
  const tags: string[] = []
  const tagRe = /#(\w+)/g
  let m: RegExpExecArray | null
  while ((m = tagRe.exec(text)) !== null) tags.push(m[1].toLowerCase())
  const ctxRe = /@(\w+)/g
  while ((m = ctxRe.exec(text)) !== null) tags.push(m[1].toLowerCase())
  return tags.length > 0 ? tags : undefined
}

function removePrefix(text: string, prefixes: string[]): string {
  let result = text.trim()
  for (const prefix of prefixes) {
    const regex = new RegExp(`^${prefix}\\s+`, 'i')
    result = result.replace(regex, '').trim()
  }
  return result
}

function extractQueryEntity(text: string): { entity?: string; timeframe?: string; intent?: string } {
  const lower = text.toLowerCase()

  const entityPatterns: Array<[RegExp, string]> = [
    [/\b(tasks?|to-do|todo|todos)\b/, 'task'],
    [/\b(habits?)\b/, 'habit'],
    [/\b(habit streak)\b/, 'habit'],
    [/\b(sleep|bedtime)\b/, 'sleep'],
    [/\b(courses?|class|classes|subject)\b/, 'course'],
    [/\b(goals?)\b/, 'goal'],
    [/\b(projects?)\b/, 'project'],
    [/\b(ideas?)\b/, 'idea'],
    [/\b(income|earnings?)\b/, 'income'],
    [/\b(time|timesheet)\b/, 'time'],
    [/\b(resources?)\b/, 'resource'],
    [/\b(opportunit(y|ies))\b/, 'opportunity'],
    [/\b(memory|memories)\b/, 'memory'],
  ]

  let entity: string | undefined
  for (const [pattern, name] of entityPatterns) {
    if (pattern.test(lower)) { entity = name; break }
  }

  const timeframePatterns: Array<[RegExp, string]> = [
    [/\b(today)\b/, 'today'],
    [/\b(tomorrow)\b/, 'tomorrow'],
    [/\b(this week)\b/, 'this_week'],
    [/\b(next week)\b/, 'next_week'],
    [/\b(this month)\b/, 'this_month'],
    [/\b(overdue|over due|past due)\b/, 'overdue'],
    [/\b(upcoming)\b/, 'upcoming'],
  ]

  let timeframe: string | undefined
  for (const [pattern, name] of timeframePatterns) {
    if (pattern.test(lower)) { timeframe = name; break }
  }

  const intentPatterns: Array<[RegExp, string]> = [
    [/\b(count|how many|total)\b/, 'count'],
    [/\b(list|show|what are|what's)\b/, 'list'],
    [/\b(summary|overview|status)\b/, 'summary'],
    [/\b(due|deadline)\b/, 'due'],
    [/\b(trend|progress|improving|declining)\b/, 'trend'],
  ]

  let intent: string | undefined
  for (const [pattern, name] of intentPatterns) {
    if (pattern.test(lower)) { intent = name; break }
  }

  return { entity, timeframe, intent }
}

const QUERY_TRIGGERS = [
  /^(what|how|when|where|why|who|show|list|display|tell\s+me|give\s+me)\b/i,
  /^(are\s+there|is\s+there|do\s+I|does\s+my)\b/i,
  /^(count|summarize|check)\b/i,
]

export function parseCommand(text: string): ParsedCommand {
  const trimmed = text.trim()
  const lower = trimmed.toLowerCase()

  if (lower.startsWith('/new task')) {
    const title = removePrefix(trimmed, ['/new task'])
    if (!title) return { type: 'unknown', confidence: 0, raw: trimmed }
    return {
      type: 'create_task',
      confidence: 0.9,
      task: {
        title,
        dueDate: extractDate(title),
        priority: extractPriority(title),
        estimatedMinutes: extractMinutes(title),
        recurrence: extractRecurrence(title),
        tags: extractTags(title),
      },
      raw: trimmed,
    }
  }

  if (lower.startsWith('/go ') || lower.startsWith('/navigate ')) {
    const dest = removePrefix(trimmed, ['/go', '/navigate'])
    return {
      type: 'navigate',
      confidence: 0.95,
      navigation: dest.toLowerCase(),
      raw: trimmed,
    }
  }

  if (lower.startsWith('/complete ') || lower.startsWith('/done ')) {
    const title = removePrefix(trimmed, ['/complete', '/done'])
    if (!title) return { type: 'unknown', confidence: 0, raw: trimmed }
    return {
      type: 'complete_task',
      confidence: 0.85,
      task: { title },
      raw: trimmed,
    }
  }

  const createPatterns = [
    /^(create|add|make)\s+(a\s+)?(task|todo|to-do)\s+(?:to\s+)?(.+)/i,
    /^remind\s+me\s+(?:to\s+)?(.+)/i,
    /^(i\s+)?need\s+to\s+(.+)/i,
    /^(schedule|plan)\s+(.+)/i,
  ]

  for (const pattern of createPatterns) {
    const match = trimmed.match(pattern)
    if (match) {
      const title = match[match.length - 1].trim()
      if (title.length < 2) continue
      return {
        type: 'create_task',
        confidence: 0.85,
        task: {
          title,
          dueDate: extractDate(trimmed),
          priority: extractPriority(trimmed),
          estimatedMinutes: extractMinutes(trimmed),
          recurrence: extractRecurrence(trimmed),
          tags: extractTags(trimmed),
        },
        raw: trimmed,
      }
    }
  }

  const donePattern = /^(mark|set)\s+(.+?)\s+(as\s+)?(done|completed|complete)/i
  const doneMatch = trimmed.match(donePattern)
  if (doneMatch) {
    return {
      type: 'complete_task',
      confidence: 0.7,
      task: { title: doneMatch[2].trim() },
      raw: trimmed,
    }
  }

  const navPatterns = [/^(go\s+to|open|show|take\s+me\s+to)\s+(.+)/i]
  for (const pattern of navPatterns) {
    const match = trimmed.match(pattern)
    if (match) {
      return {
        type: 'navigate',
        confidence: 0.8,
        navigation: match[2].trim().toLowerCase(),
        raw: trimmed,
      }
    }
  }

  for (const trigger of QUERY_TRIGGERS) {
    if (trigger.test(trimmed)) {
      return {
        type: 'query',
        confidence: 0.65,
        query: extractQueryEntity(trimmed),
        raw: trimmed,
      }
    }
  }

  return { type: 'unknown', confidence: 0.2, raw: trimmed }
}

const ROUTE_ALIASES: Record<string, string> = {
  dashboard: '/dashboard',
  tasks: '/dashboard/tasks',
  'task list': '/dashboard/tasks',
  habits: '/dashboard/habits',
  sleep: '/dashboard/sleep',
  courses: '/dashboard/courses',
  goals: '/dashboard/goals',
  chat: '/dashboard/chat',
  projects: '/dashboard/projects',
  ideas: '/dashboard/ideas',
  'time tracking': '/dashboard/time',
  time: '/dashboard/time',
  income: '/dashboard/income',
  resources: '/dashboard/resources',
  opportunities: '/dashboard/opportunities',
  memory: '/dashboard/memory',
  knowledge: '/dashboard/knowledge',
  'knowledge graph': '/dashboard/knowledge',
  roadmap: '/dashboard/roadmap',
  settings: '/dashboard/settings',
  analytics: '/dashboard/analytics',
  'youtube vault': '/dashboard/youtube-vault',
  'focus mode': '/dashboard/focus',
  automation: '/dashboard/automation',
  review: '/dashboard/review',
  briefing: '/dashboard/briefing',
  academics: '/dashboard/academics',
  'daily briefing': '/dashboard/briefing',
  'weekly review': '/dashboard/review',
}

export function resolveNavigation(nav: string): string | undefined {
  const key = nav.toLowerCase().trim()
  if (ROUTE_ALIASES[key]) return ROUTE_ALIASES[key]
  for (const [alias, route] of Object.entries(ROUTE_ALIASES)) {
    if (key.includes(alias) || alias.includes(key)) return route
  }
  return undefined
}

File: apps/api/app/api/nlp.py
import re
from datetime import datetime, timedelta, timezone
from fastapi import APIRouter, Depends, HTTPException
from config.core.supabase import get_supabase_client
from config.core.auth import get_current_user
from shared.utils.logger import logger
from database.schemas.nlp import NLPParseRequest, NLPParseResponse

router = APIRouter()

DAY_NAMES = {
    "sunday": 0,
    "monday": 1,
    "tuesday": 2,
    "wednesday": 3,
    "thursday": 4,
    "friday": 5,
    "saturday": 6,
    "sun": 0,
    "mon": 1,
    "tue": 2,
    "wed": 3,
    "thu": 4,
    "fri": 5,
    "sat": 6,
}

ROUTE_ALIASES = {
    "dashboard": "/dashboard",
    "tasks": "/dashboard/tasks",
    "habits": "/dashboard/habits",
    "sleep": "/dashboard/sleep",
    "courses": "/dashboard/courses",
    "goals": "/dashboard/goals",
    "chat": "/dashboard/chat",
    "projects": "/dashboard/projects",
    "ideas": "/dashboard/ideas",
    "time": "/dashboard/time",
    "income": "/dashboard/income",
    "resources": "/dashboard/resources",
    "opportunities": "/dashboard/opportunities",
    "memory": "/dashboard/memory",
    "knowledge": "/dashboard/knowledge",
    "roadmap": "/dashboard/roadmap",
    "settings": "/dashboard/settings",
    "analytics": "/dashboard/analytics",
    "youtube vault": "/dashboard/youtube-vault",
    "focus": "/dashboard/focus",
    "automation": "/dashboard/automation",
    "review": "/dashboard/review",
    "briefing": "/dashboard/briefing",
    "academics": "/dashboard/academics",
}


def extract_date(text: str) -> str | None:
    now = datetime.now(timezone.utc)
    today = now.strftime("%Y-%m-%d")
    lower = text.lower()

    if "today" in lower:
        return today
    if "tomorrow" in lower:
        return (now + timedelta(days=1)).strftime("%Y-%m-%d")

    for name, idx in DAY_NAMES.items():
        if name in lower:
            target = now + timedelta(days=(idx - now.weekday() + 7) % 7)
            return target.strftime("%Y-%m-%d")

    match = re.search(r"(\d{1,2})/(\d{1,2})(?:/(\d{2,4}))?", text)
    if match:
        m, d, y = match.groups()
        year = y if y else str(now.year)
        if len(year) == 2:
            year = f"20{year}"
        return f"{year}-{m.zfill(2)}-{d.zfill(2)}"

    rel = re.search(r"in\s+(\d+)\s+(day|days|week|weeks)", lower)
    if rel:
        num = int(rel.group(1))
        unit = rel.group(2).lower()
        target = now + timedelta(days=num * 7) if unit.startswith("week") else now + timedelta(days=num)
        return target.strftime("%Y-%m-%d")

    return None


def extract_priority(text: str) -> str | None:
    lower = text.lower()
    if re.search(r"\b(urgent|critical|asap|high priority|important)\b", lower):
        return "high"
    if re.search(r"\b(low priority|whenever|someday|optional)\b", lower):
        return "low"
    return None


def extract_minutes(text: str) -> int | None:
    match = re.search(r"(\d+)\s*(mins|min|minute|minutes|hour|hours|hr|hrs)\b", text.lower())
    if not match:
        return None
    num = int(match.group(1))
    unit = match.group(2).lower()
    return num * 60 if unit.startswith("hour") or unit.startswith("hr") else num


def resolve_route(nav: str) -> str | None:
    key = nav.strip().lower()
    if key in ROUTE_ALIASES:
        return ROUTE_ALIASES[key]
    for alias, route in ROUTE_ALIASES.items():
        if key in alias or alias in key:
            return route
    return None


@router.post("/parse", summary="Parse natural language input", response_model=NLPParseResponse)
async def parse_natural_language(req: NLPParseRequest, current_user=Depends(get_current_user)):
    text = req.text.strip()
    lower = text.lower()

    # Slash command: /new task ...
    if lower.startswith("/new task "):
        title = re.sub(r"^/new task\s+", "", text, flags=re.IGNORECASE).strip()
        if title:
            return NLPParseResponse(
                type="create_task",
                confidence=0.9,
                task={
                    "title": title,
                    "due_date": extract_date(title),
                    "priority": extract_priority(title),
                    "estimated_minutes": extract_minutes(title),
                },
                raw=text,
            )

    if lower.startswith("/go ") or lower.startswith("/navigate "):
        dest = re.sub(r"^/(go|navigate)\s+", "", text, flags=re.IGNORECASE).strip()
        route = resolve_route(dest)
        if route:
            return NLPParseResponse(type="navigate", confidence=0.95, navigation=route, raw=text)

    # Natural language patterns
    patterns = [
        (r"^(create|add|make)\s+(a\s+)?(task|todo)\s+(?:to\s+)?(.+)", "create_task"),
        (r"^remind\s+me\s+(?:to\s+)?(.+)", "create_task"),
        (r"^(i\s+)?need\s+to\s+(.+)", "create_task"),
        (r"^(schedule|plan)\s+(.+)", "create_task"),
    ]

    for pattern, cmd_type in patterns:
        match = re.search(pattern, text, re.IGNORECASE)
        if match:
            title = match.group(match.lastindex).strip()
            if len(title) >= 2:
                return NLPParseResponse(
                    type=cmd_type,
                    confidence=0.85,
                    task={
                        "title": title,
                        "due_date": extract_date(text),
                        "priority": extract_priority(text),
                        "estimated_minutes": extract_minutes(text),
                    },
                    raw=text,
                )

    # Navigation
    nav_match = re.search(r"^(go\s+to|open|show|take\s+me\s+to)\s+(.+)", text, re.IGNORECASE)
    if nav_match:
        dest = nav_match.group(2).strip().lower()
        route = resolve_route(dest)
        if route:
            return NLPParseResponse(type="navigate", confidence=0.8, navigation=route, raw=text)

    return NLPParseResponse(type="unknown", confidence=0.2, raw=text)


@router.post("/execute", summary="Execute NL command", response_model=dict)
async def execute_command(req: dict, current_user=Depends(get_current_user)):
    cmd_type = req.get("type")
    task_data = req.get("task") or {}

    if cmd_type == "create_task":
        supabase = get_supabase_client()
        from uuid import uuid4

        now = datetime.now(timezone.utc).isoformat()
        record = {
            "id": str(uuid4()),
            "user_id": current_user.user.id,
            "title": task_data.get("title", "Untitled"),
            "status": "pending",
            "priority": task_data.get("priority", "medium"),
            "due_date": task_data.get("due_date"),
            "estimated_minutes": task_data.get("estimated_minutes"),
            "created_at": now,
        }
        try:
            supabase.table("tasks").insert(record).execute()
            return {"success": True, "message": f'Task "{record["title"]}" created', "data": record}
        except Exception as e:
            logger.error("Failed to create task via NL", error=str(e))
            raise HTTPException(status_code=500, detail="Failed to create task")
    elif cmd_type == "navigate":
        return {"success": True, "redirect_url": req.get("navigation", "/dashboard")}
    elif cmd_type == "schedule":
        supabase = get_supabase_client()
        from uuid import uuid4

        now = datetime.now(timezone.utc).isoformat()
        sched = req.get("schedule") or {}
        record = {
            "id": str(uuid4()),
            "user_id": current_user.user.id,
            "start_time": sched.get("start_time", now),
            "duration_minutes": sched.get("duration_minutes", 30),
            "description": sched.get("description", "Scheduled task"),
            "category": sched.get("category", "focus"),
            "created_at": now,
        }
        try:
            supabase.table("time_entries").insert(record).execute()
            return {
                "success": True,
                "message": f'Scheduled "{record["description"]}" for {record["duration_minutes"]}min',
            }
        except Exception as e:
            logger.error("Failed to create time entry via NL", error=str(e))
            raise HTTPException(status_code=500, detail="Failed to schedule")
    else:
        return {"success": False, "message": "Unknown command type"}
